### ISSUE TITLE
cam_dev is unable to achieve bit-for-bit restarts in numerous configurations

### ISSUE THREAD

**Issue by cacraigucar** [ESCOMP/CAM#655]
### What happened?

The initial problem of non-bit-for-bit restarts was documented when running ERI_D_Ln18_Vnuopc.ne30pg3_ne30pg3_mg17.F2000dev.cheyenne_intel.cam-outfrq9s

An identical test using F2000climo instead of F2000dev had bit-for-bit restarts

After further testing, it was determined that this test also failed with an FV grid:  ERI_D_Ln18_Vnuopc.f19_f19_mg17.F2000dev.cheyenne_intel.cam-outfrq9s

The same behavior was seen in ERP tests and the very coarse FV grid f10_f10_mg37.  This grid was used extensively to determine a pattern.  The pattern from ERP_D_LnXX_Vnuopc.f10_f10_mg37.F2000dev.cheyenne_intel.cam-outfrq9s is as follows:

(X=fail bit-for-bit, P=passed bit-for-bit and the number is the number of timesteps used in the LnXX testing constraint)

  4....5....6....7....8....9....10....11....12....13....14....15....16....17....18
 X....**P**....X...X....**P....P**.....X......X.......**P......P**......X......X.......**P......P**......X

### What are the steps to reproduce the bug?

./create_test ERP_D_LnXX_Vnuopc.f10_f10_mg37.F2000dev.cheyenne_intel.cam-outfrq9s where XX is the number of timesteps which wish to use

### What CAM tag were you using?

cam6_3_077

### What machine were you running CAM on?

CISL machine (e.g. cheyenne)

### What compiler were you using?

Intel

### Will you be addressing this bug yourself?

No

### Extra info

@adamrher - Does this pattern perhaps indicate anything in cam_dev which might be the culprit?

@bstephens82 - Please augment this issue with observations which you have that I may not have already covered.

**Comment by bstephens82** [comment]
It may be helpful to augment your pass/fail table to include the timestep from which the restart tests restart, which you told me is $STOP_N/2+1, where $STOP_N/2 rounds down if STOP_N is odd.  This would give:
| STOP_N | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| restart timestep | 3 | 3 | 4 | 4 | 5 | 5 | 6 | 6 | 7 | 7 | 8 | 8 | 9 | 9 | 10 |
| P/X | X | P | X | X | P | P | X | X | P | P | X| X| P | P | X |

With the exception of STOP_N=4 (could this be exceptional in some way because of how short it is?), this pattern suggests that starting from odd timesteps will give BFB restarts but starting from even timesteps will not.

**Comment by adamrher** [comment]
can you check whether ```nextsw_cday``` is in the restart files? An every other time-step suggests to me problems with restarting radiation. I think maybe Ben came to the conclusion independently (you asked me for the radiation time-stepping variables the other day).

**Comment by bstephens82** [comment]
> can you check whether `nextsw_cday` is in the restart files?

Yes I find that `nextsw_cday` is present in my cam_dev restart files.

**Comment by cacraigucar** [comment]
Looking at STOP_N=5 through 11, the nextsw_cday =1 for the tests which fail and -1 for the tests which pass.

The outlier STOP_N which fails also has nextsw_day=-1 which is contrary to the pattern reported above.

So the augmented table is:
| STOP_N | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| restart timestep | 3 | 3 | 4 | 4 | 5 | 5 | 6 | 6 | 7 | 7 | 8 | 8 | 9 | 9 | 10 |
| P/X | X | P | X | X | P | P | X | X | P | P | X| X| P | P | X |
|nextsw_cday| -1|-1|1|1|-1|-1|1|1|

**Comment by adamrher** [comment]
So this is consistent with the ERP tests passing, right? In those tests, I believe the restart timestep is nstep=5, and runs to nstep=9. Since 5 is odd, it passes.

I wonder if we set the radiation time-step to run every time-step (which will force nextsw_cday = 1 for all time) is it will fail no matter whether even or odd time-step (apologies I won't be able to start digging into this myself until later ... I'm just being the peanut gallery here).

**Comment by bstephens82** [comment]
> So this is consistent with the ERP tests passing, right? In those tests, I believe the restart timestep is nstep=5, and runs to nstep=9. Since 5 is odd, it passes.
> 
> I wonder if we set the radiation time-step to run every time-step (which will force nextsw_cday = 1 for all time) is it will fail no matter whether even or odd time-step (apologies I won't be able to start digging into this myself until later ... I'm just being the peanut gallery here).

I ran cam_dev with radiation called every timestep and it did fail.  But I didn't do the kind of extensive testing Cheryl has done.

**Comment by JulioTBacmeister** [comment]
Question to Adam: Do you expect restarts to fail ALWAYS when radiation is called every timestep?

**Comment by cacraigucar** [comment]
I will run setting iradsw=1 and iradlw=1 for STOP_N=7,8 and 9 and see if they all fail restarts.

**Comment by adamrher** [comment]
Julio - not until today ;) I had made code mods that were intended to permit restarts every time-step, including the even time-steps which seem to be the problem here ... but since our restart tests always restart on an odd time-step, we don't actually test for the this and so this seems to have slipped through the cracks.

caveat, we're yet to really prove it's the radiation ... cheryl tests should tell us

**Comment by cacraigucar** [comment]
For STOP_N=7, 8, 9 using iradsw=1 and iradlw=1, all three runs fail bit-for-bit baselines for restarts and nextsw_cday=1

**Comment by adamrher** [comment]
thx cheryl, that's fairly strong evidence that this is related to changes I made to radiation in cam_dev. let me look into this a bit later and Ill update this thread.

**Comment by adamrher** [comment]
I've not gotten too far today on this issue. I've isolated the answer changes to the aerosols (e.g., bc_a1), chemical species (e.g., SO2) and radiative fluxes (e.g., FSDS). This is clear when you do a cprnc to nstep=7 in the base and rest cases (rest begins at nstep=6, but that time-step is bfb w/ the base).

**Comment by adamrher** [comment]
@fvitt I am having trouble finding the outfld calls for the aerosol/chemical species tracers. It depends on the configuration, but for F2000, it is clear that the aerosol / chemistry modules are declaring these tracers and are responsible for the outfld calls. So my question is this: which subroutine in tphysbc/tphysac contain the outfld calls for aerosols (e.g., bc_a1)? Which subroutine contains the outfld calls for chemical species (e.g., SO2)?

My guess is that both aerosols and chemical species outfld calls are contained in the ```call chem_timestep_tend``` in the ```t_startf('adv_tracer_src_snk')``` code block of tphysac. No?

**Comment by adamrher** [comment]
I've further trapped this bug. The fist occurrence of non-bfb is in the shortwave radiative fluxes (e.g., FSDS, FSNS, etc ...). No problems with the longwave. The optical depths associated with aerosols are bfb, so there's no problem w/ the aerosols (phew!). Rather, the surface albedo appears to be the culprit as it's the only non-bfb var in the clm output at the time-step where non-bfb begins. 

I suspect that since radiation is downstream of the CPL in cam_dev, it is using the albedo from the CPL instead of the albedo from the restarts. But they should be identical so that's a problem.

**Comment by bstephens82** [comment]
Today I came across a few things that might be of interest, but no smoking gun.

One thing I found that may be of concern is that after adding four albedo variables to user_nl_clm (ALBI, ALBD, ALBGRD, and ALBGRI), all of the restart tests I did failed, including not just all F2000dev tests but also F2000climo tests.  All of the unexpected failures occurred because ALBGRD and ALBGRI did not match after the restart test was completed, but if you remove these variables from user_nl_clm, the tests pass again.  I did not spend time trying to track this down, since it seemed like these might be diagnostic variables that can be added to the history files but that do not impact BFB restarts.  But, in the little time I did spend looking at this, I did come across a line of code in the land model that looked suspicious to me.  This is in CAM/components/clm/src/biogeophys/SurfaceAlbedoType.F90, around line 190 or so:
~~~
    this%coszen_col(begc:endc) = spval
    call hist_addfld1d (fname='COSZEN', units='none', &
         avgflag='A', long_name='cosine of solar zenith angle', &
         ptr_col=this%coszen_col, default='inactive')

    this%albgri_col(begc:endc,:) = spval          <========= should this line start with this%albgrd_col?
    call hist_addfld2d (fname='ALBGRD', units='proportion', type2d='numrad', &
         avgflag='A', long_name='ground albedo (direct)', &
         ptr_col=this%albgrd_col, default='inactive')

    this%albgri_col(begc:endc,:) = spval
    call hist_addfld2d (fname='ALBGRI', units='proportion', type2d='numrad', &
         avgflag='A', long_name='ground albedo (indirect)', &
         ptr_col=this%albgri_col, default='inactive')
~~~
I don't know this code well enough to be sure this is a bug but it looked odd to me.


Based on print statements I wasn't able to find any obvious differences between base and restart regarding radiation---it looks like the calls to radiation are happening at the same places in the two experiments, etc.  Based on other print statements, however, it does tentatively look to me like the SW radiation subroutine is using a different albedo variable in the base vs restart after the restart time step, lending weight to Adam's idea that radiation might be using albedo from the coupler rather than from the restarts, so I will continue to pursue that idea.

**Comment by ekluzek** [comment]
I think the comment above refers to the following issue in CTSM.

https://github.com/ESCOMP/CTSM/issues/1788

This points out that this is a bigger problem than it appears since ALBGRD is not normally output.

**Comment by bstephens82** [comment]
> I think the comment above refers to the following issue in CTSM.
> 
> [ESCOMP/CTSM#1788](https://github.com/ESCOMP/CTSM/issues/1788)
> 
> This points out that this is a bigger problem than it appears since ALBGRD is not normally output.

Hi @ekluzek, just to be clear, I did try correcting this line of code and it didn't make ALBGRD or ALBGRI match for the restart tests I was doing, i.e. including those two variables in user_nl_clm still caused restart tests to fail that otherwise would pass.

**Comment by klindsay28** [comment]
I think that this issue is impacting coupled experiments with cam_dve that @cecilehannay is running.
For separate work, I cloned one of her experiments, and my solutions diverge from hers after an earlier continuation in my clone.
In investigating the divergence, I noticed that CICE writes out the value of `nextsw_cday`.
In a continuous run, the written values around 0001-01-03 00000 are
```
0.29791666666667D+01
-0.10000000000000D+01
0.30208333333333D+01
-0.10000000000000D+01
0.30625000000000D+01
```

With a continuation at 0001-01-03 00000, the values are
```
0.29791666666667D+01
-0.10000000000000D+01
0.30000000000000D+01
-0.10000000000000D+01
0.30625000000000D+01
```

The first non `-1` value after the continuation is 3.0, instead of the 3.0208333 value in the continuous run.

So it appears that `nextsw_cday` is not getting sent appropriately immediately after a continuation.

**Comment by bstephens82** [comment]
@klindsay28 thanks I was actually just about to report that this seems to be the problem--nextsw_cday being saved as an integer rather than as a real number. This seems to impact the land model's calculation of coszen. I haven't confirmed that correcting this gives BFB restarts but I'm hoping so and should have confirmation soon.

**Comment by klindsay28** [comment]
That would explain the change in `nextsw_cday` from `3.0208333` to `3.0`. Good catch.

**Comment by bstephens82** [comment]
Happily I can confirm that nextsw_cday being saved as an integer does seem to be the problem :)  
The cam_dev restart test PASSES when I make two changes in src/physics/rrtmg/radiation.F90:
1. Include pio_double in the use statement for pio_types:
~~~
use pio,                 only: file_desc_t, var_desc_t,               &
                               pio_int, pio_double, pio_noerr,                    &
~~~
and 2. change the type of nextsw_cday from pio_int to pio_double:
~~~
   ierr = pio_def_var(File, 'nextsw_cday', pio_double, nextsw_cday_desc)
~~~

**Comment by ekluzek** [comment]
@bstephens82 good job on tracking this down! 

Can you also look into when nextsw_cday was changed to be an integer rather than a double? It's double in CTSM and should have always been a double.It would be good to figure out when (and why) it changed to prevent such a change in the future. It would also be good to add a test that does the restart on the non-radiation time-step so that we could catch such a problem in the future. This is perhaps more of a task for @cacraigucar .

**Comment by JulioTBacmeister** [comment]
Thanks Ben! Excellent!

**Comment by bstephens82** [comment]
> @bstephens82 good job on tracking this down!
> 
> Can you also look into when nextsw_cday was changed to be an integer rather than a double? It's double in CTSM and should have always been a double.It would be good to figure out when (and why) it changed to prevent such a change in the future. It would also be good to add a test that does the restart on the non-radiation time-step so that we could catch such a problem in the future. This is perhaps more of a task for @cacraigucar .

Looks like it was added to the restart files as an integer at the beginning of this year: https://github.com/ESCOMP/CAM/commit/c554aa6abb797d346f4f17997c9ac2d218382368, but I'm not sure if it was intentionally an integer or not.

**Comment by gold2718** [comment]
> Can you also look into when nextsw_cday was changed to be an integer rather than a double?

The answer is never. It turns out that before cam_dev, CAM did not need to save `nextsw_cday`. When @adamrher added it to the restart, he fell prey to Murphy's law and copied and pasted from the wrong example. It also explains why it works sometimes and not others (should work when `int(nextsw_cday) == nextsw_cday`.

**Comment by klindsay28** [comment]
I confirm that the fix of @bstephens82 leads to exact restarts in the coupled experiment that I'm running.
Thanks @bstephens82!

**Comment by adamrher** [comment]
Thanks Ben and Keith for help tracking down this bug. @cacraigucar @gold2718 should @bstephens82 make a PR? 

@bstephens82 you will need to make these identical changes to src/physics/camrt/radiation.F90 as well, so we don't break CAM4 physics (easily seen by clicking the hash you provided that points to mods I made for putting nextsw_cday in the restarts).

**Comment by cacraigucar** [comment]
@bstephens82 - If you want to make a PR or point me to your changes, I think I'll be making a tag ASAP with this change and the changes needed to address #631

**Comment by cacraigucar** [comment]
I hate the be the bearer of bad news, but I ran Ln4, Ln5, Ln6 and Ln7.  5,6 and 7 are all bit-for-bit for restarts, but Ln4 is not.  Since this didn't match the pattern originally with failing/passing restart tests, I suspected it might not be fixed.  @adamrher - is there some special logic at the beginning of the run which could trip this up?  I am using the fixes that @bstephens82 provided.

**Comment by gold2718** [comment]
Is Ln4 supposed to work? Which test type is this? Most restart tests have a minimum length they work with but most are not great at checking and producing an error if your run is too short.

**Comment by adamrher** [comment]
looking at the tables near the top of this thread Ln4, Ln5 both use the restart files from nstep=3. Without even thinking too deeply about the cause for the Ln4 fail, I am willing to bet the buffer isn't fully populated for a bfb restart for at least a time-step past the nstep=0 (0.5?) initialization phase. I wouldn't be surprised if cam_dev pushed that out another time-step ... that's my wild guess.

**Comment by bstephens82** [comment]
I don't know about the Ln4 test, but I thought I'd add that today I tried some longer tests, 5-6 months, and they also restarted BFB with the fix.

**Comment by ekluzek** [comment]
I think the minimum length in units for a restart test is "5" with "10" being the default. The unit is the unit length (in this case steps, but can be any unit like, days, months or years) I think that's because as @adamrher points out the restart is done 3/5th's into the whole time period. So I think it's likely that the 4-step test just doesn't work because it's too short. It does sometime tell you it's too short, but maybe not all the time.

**Comment by adamrher** [comment]
Thanks, Erik. I think that settles this, that the Ln4 fail is not a show-stopper?

**Comment by cacraigucar** [comment]
I agree that Ln4 can be ignored based on the feedback from others on this most likely being an unsupported test.
